rfd/0209-mcp-access.md

```diff
@@ -0,0 +1,510 @@
+or authorization. OAuth authorization was recently added to the specification,
+but as of the time of writing, it is still new and not widely adopted.
+
+With Teleport's support for MCP servers, users can:
+- Access MCP servers on remote machines where transport is secured with TLS
+- Control access to specific MCP servers and specific tools defined by Teleport roles
+- Track activity through Teleport's audit log
+
+## Scope
+
+Teleport will support stdio-based MCP servers in the initial implementation,
+with [streamable
+HTTP](https://modelcontextprotocol.io/specification/2025-03-26/basic/transports#streamable-http))
+and
+[OAuth](https://modelcontextprotocol.io/specification/2025-03-26/basic/authorization)
+support planned for future iterations (see the "Future development" section for
+details).
+
+## Details
+
+To expedite the initial rollout, the first iteration will focus on adding MCP
+protocol support for Application Access, similar to existing HTTP and TCP
+applications.
+
+### UX
+
+#### editor - configure a filesystem MCP server and define RBAC
+To configure a [filesystem MCP
+server](https://github.com/modelcontextprotocol/servers/tree/main/src/filesystem)
+with `npx` and a [Slack API MCP
+server](https://github.com/modelcontextprotocol/servers/tree/main/src/slack)
+with `docker` via Teleport App service:
+```yaml
+app_service:
+  enabled: true
+  apps:
+  - name: "dev-files"
+    description: "Shared files for developers"
+    labels:
+      env: "dev"
+    mcp:
+      # Command and arguments to launch a stdio-based MCP server per client connection
+      # on the Teleport service host.
+      command: "npx"
+      args: ["-y", "@modelcontextprotocol/server-filesystem", "/home/dev/files"]
+      # Specifies the local user account under which the command will be executed.
+      # Required for stdio-based MCP servers.
+      run_as_local_user: "dev"
+
+  - name: "dev-slack-api"
+    description: "Slack API to dev channels"
+    labels:
+      env: "dev"
+    mcp:
+      # Recommend containers for launching stdio-based MCP servers. Use an env file
+      # for passing secrets.
+      command: "docker"
+      args: ["run", "--rm", "--env-file", "/etc/teleport/dev_slack_api_env.list", "-i", "mcp/slack"]
+      run_as_local_user: "docker"
+```
+
+To create a Teleport role for developers that can access only dev MCP servers,
+have read-only filesystem access, and no permission to post Slack messages:
+```yaml
+kind: role
+version: v8
+metadata:
+  name: dev
+spec:
+  allow:
+    # Labels of the MCP servers to allow access to.
+    app_labels:
+      "env": "dev"
+    mcp:
+      # The name of the MCP tools to allow access to.
+      # The wildcard character '*' matches any sequence of characters.
+      # If the value begins with '^' and ends with '$', it is treated as a regular expression.
+      # This value field also supports variable interpolation.
+      # No tools are allowed if not specified.
+      tools:
```

Comment:
is this called tools or services in MCP spec?

Reply:
[Tools](https://modelcontextprotocol.io/specification/2025-03-26/server/tools)